### PR TITLE
WIP: fixed typescript definitions 

### PR DESCRIPTION
**Pull request by thomschke**
#### What does this PR do?
Providing some bugfixes/addons for typescript definitions.

#### What are the relevant issues?
#2203

### CODE REVIEW

**Comment by alansouzati** [comment]
thanks for your contribution. Just keep in mind that most of the typescript definitions are automatically generated based on the `doc` module in each component folder. I'm fine with your contributions, I added one comment. In the meantime I will check why the generator did not get the array type properly for Accordion.

File: src/js/components/Accordion/index.d.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import * as React from "react";
 export interface AccordionProps {
   activeIndex?: number | number[];
   animate?: boolean;
-  children: React.ReactNode[];
```
Comment (alansouzati) [inline src/js/components/Accordion/index.d.ts]: curious why we are removing the array here?

**Comment by alansouzati** [comment]
actually you are removing the array from Accordion typescript definitions. maybe we should augment the children to be either?

**Comment by alansouzati** [comment]
here is an example of how to change Tabs and Accordion to accept both types.

https://github.com/grommet/grommet/pull/2255

**Comment by thomschke** [comment]
Actual I get following compile error when importing the declarations:

```
JSX element type 'Component<...>' is not a constructor function for JSX elements.
  Types of property 'render' are incompatible.
    Type '() => ReactNode' is not assignable to type '{ (): ReactNode; (): false | Element | null; }'.
      Type 'ReactNode' is not assignable to type 'false | Element | null'.
        Type 'undefined' is not assignable to type 'false | Element | null'.

```
see DefinitelyTyped/DefinitelyTyped#28636 

You can extend `React.SFC` (instead of `React.ComponentType`) as a workaround.

**Comment by thomschke** [comment]
ReactNode -> ReactFragment -> ReactNodeArray [(see)](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2b6e08299d0793602eae13f5ea1a77fe18981c13/types/react/index.d.ts#L150-L152)

**Comment by alansouzati** [comment]
perfect @thomschke I did not know about this. so, do you mind updating your PR to not change the typescript definition directly but do inside the doc module itself? you can use my other PR as a reference, then I can close it later on.

For the Accordion we can update the [doc](https://github.com/grommet/grommet/blob/NEXT/src/js/components/Accordion/doc.js#L31) to be this:

```
    children: PropTypes.node.description(
      'Array of AccordionPanels.'
    ).isRequired
```

Let me know your thoughts here

**Comment by thomschke** [comment]
Yes, I think you can use `PropTypes.node` [(see)](https://github.com/facebook/prop-types/blob/cfc7f43af3d74978ec82052fe6bb8212f09daca0/README.md#usage):

>   // Anything that can be rendered: numbers, strings, elements or **an array**
>   // (or fragment) containing these types.
>   optionalNode: PropTypes.node

**Comment by thomschke** [comment]
I closed PR DefinitelyTyped/DefinitelyTyped#28636.

But I've found that you're not exporting the component directly, but instead exporting a wrapper using compose. 

This is a HOC ! So you should generate declarations based on `React.StatelessComponent` ( instead on [`React.ComponentType`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2b6e08299d0793602eae13f5ea1a77fe18981c13/types/react/index.d.ts#L53) which includes `React.ComponentClass`), right? 

If done, then there are no compile errors anymore.

**Comment by thomschke** [comment]
Due to changing `tools/generate-readme-ts.js` I have to merge your changes from #2255 to generate correct declaration files.

So there will be merge conflicts if you merge both PRs, sorry.

**Comment by thomschke** [comment]
BTW. react-desc@3.5.0 is not published yet

**Comment by alansouzati** [comment]
beautifully done. thanks for your contribution. welcome to the team

**Comment by alansouzati** [comment]
let me push the changes for react-desc, sorry I forgot to do it.

**Pull request merged**